Account wide Reagent Bank/Bag

jacksonwilliams opened the pull request:
<!-- First of all, THANK YOU for your contribution. -->

## Changes Proposed:
-  Update reagent store / pouch to be used account-wide


## Tests Performed:

- New account created.
- New character created.
- character teleported to bootybay
- added 1000g to character
- added 100 silverleaf to character
- added 100 peacebloom to character
- purchased reagent bag
- herbs deposited into reagent bag
- verified herbs visible in reagent bag

- migration sql run on database.
- server restarted

- new character created
- character teleported to bootybay
- interact with winzig
- check herb tab of reagent bank
- confirmed 100 silverleaf and 100 peacebloom was visable.

## How to Test the Changes:

1. goto winzig on character 1 / purchase reagent pouch
2. deposit items into the reagent bank
3. can items be seen on character 2

## Known Issues and TODO List:
<!-- Is there anything else left to do after this PR? -->

- None

jarrodjeffi commented:
We'd need some sort of migration script from `character_id` to `account_id` but this looks good.

jarrodjeffi commented on modules/mod-winzig/src/WinzigCreature.cpp:
```diff
@@ -37,8 +37,9 @@ std::string npc_winzig::GetItemIcon(uint32 entry, uint32 width, uint32 height, i
     // This query can be changed to async to improve performance, but there will be some visual bugs because the query will not be done executing when the menu refreshes
-    std::string query = "SELECT amount FROM reagent_bank WHERE character_id = " + std::to_string(player->GetGUID().GetCounter()) + " AND item_entry = " + std::to_string(entry);
-    QueryResult result = CharacterDatabase.Query("SELECT amount FROM reagent_bank WHERE character_id = " + std::to_string(player->GetGUID().GetCounter()) + " AND item_entry = " + std::to_string(entry));
+    // update to be account bound: player->GetSession()->GetAccountId()
```
This comment is not necessary.

jacksonwilliams replied:
removed.

jacksonwilliams commented:
> We'd need some sort of migration script from `character_id` to `account_id` but this looks good.

I've written up an SQL file that should migrate the data, but I would like to test it before I commit it.

jacksonwilliams commented:
Tested.

New account created.
New character created.
character teleported to bootybay
added 1000g to character
added 100 silverleaf to character
added 100 peacebloom  to character

purchased reagent bag
herbs deposited into reagent bag

migration sql run on database.
server restarted

new character created
character teleported to bootybay
interact with winzig
check herb tab of reagent bank
confirmed 100 silverleaf and 100 peacebloom was visable.

jarrodjeffi commented on modules/mod-winzig/src/WinzigCreature.cpp:
```diff
@@ -177,6 +177,7 @@ bool npc_winzig::OnGossipHello(Player* player, Creature* creature)
     AddGossipItemFor(player, GOSSIP_ICON_MONEY_BAG, GetItemIcon(2604, 30, 30, -18, 0) + "Other Trade Goods", ITEM_SUBCLASS_TRADE_GOODS_OTHER, 0);
     AddGossipItemFor(player, GOSSIP_ICON_MONEY_BAG, GetItemIcon(38682, 30, 30, -18, 0) + "Armor Vellum", ITEM_SUBCLASS_ARMOR_ENCHANTMENT, 0);
     AddGossipItemFor(player, GOSSIP_ICON_MONEY_BAG, GetItemIcon(39349, 30, 30, -18, 0) + "Weapon Vellum", ITEM_SUBCLASS_WEAPON_ENCHANTMENT, 0);
+    AddGossipItemFor(player, GOSSIP_ICON_MONEY_BAG, GetItemIcon(2724, 30, 30, -18, 0) + "Quest Items", ITEM_SUBCLASS_QUEST, 0);
```
Did you add this?

jacksonwilliams replied:
removed.
